src/analysis/calculus/iterated_deriv.lean

```diff
@@ -0,0 +1,286 @@
+  { simpa [iterated_fderiv_within_eq_equiv_comp, continuous_linear_equiv.comp_differentiable_on_iff] }
+end
+
+/-- To check that a function is `n` times continuously differentiable, it suffices to check that its
+first `n` derivatives are differentiable. This is slightly too strong as the condition we
+require on the `n`-th derivative is differentiability instead of continuity, but it has the
+advantage of avoiding the discussion of continuity in the proof (and for `n = ∞` this is optimal).
+-/
+lemma times_cont_diff_on_of_differentiable_on_deriv {n : with_top ℕ}
+  (h : ∀(m : ℕ), (m : with_top ℕ) ≤ n → differentiable_on 𝕜 (iterated_deriv_within m f s) s) :
+  times_cont_diff_on 𝕜 n f s :=
+begin
+  apply times_cont_diff_on_of_differentiable_on,
+  simpa [iterated_fderiv_within_eq_equiv_comp,
+    continuous_linear_equiv.comp_differentiable_on_iff, -coe_fn_coe_base],
+end
+
+/-- On a set with unique derivatives, a `C^n` function has derivatives up to `n` which are
+continuous. -/
+lemma times_cont_diff_on.continuous_on_iterated_deriv_within {n : with_top ℕ} {m : ℕ}
+  (h : times_cont_diff_on 𝕜 n f s) (hmn : (m : with_top ℕ) ≤ n) (hs : unique_diff_on 𝕜 s) :
+  continuous_on (iterated_deriv_within m f s) s :=
+begin
+  simp [iterated_deriv_within_eq_equiv_comp, continuous_linear_equiv.comp_continuous_on_iff,
+    -coe_fn_coe_base],
+  exact h.continuous_on_iterated_fderiv_within hmn hs
+end
+
+/-- On a set with unique derivatives, a `C^n` function has derivatives less than `n` which are
+differentiable. -/
+lemma times_cont_diff_on.differentiable_on_iterated_deriv_within {n : with_top ℕ} {m : ℕ}
+  (h : times_cont_diff_on 𝕜 n f s) (hmn : (m : with_top ℕ) < n) (hs : unique_diff_on 𝕜 s) :
+  differentiable_on 𝕜 (iterated_deriv_within m f s) s :=
+begin
+  simp [iterated_deriv_within_eq_equiv_comp, continuous_linear_equiv.comp_differentiable_on_iff,
+    -coe_fn_coe_base],
+  exact h.differentiable_on_iterated_fderiv_within hmn hs
+end
+
+/-- The property of being `C^n`, initially defined in terms of the Fréchet derivative, can be
+reformulated in terms of the one-dimensional derivative on sets with unique derivatives. -/
+lemma times_cont_diff_on_iff_continuous_on_differentiable_on_deriv {n : with_top ℕ}
+  (hs : unique_diff_on 𝕜 s) :
+  times_cont_diff_on 𝕜 n f s ↔
+  (∀m:ℕ, (m : with_top ℕ) ≤ n → continuous_on (iterated_deriv_within m f s) s)
+  ∧ (∀m:ℕ, (m : with_top ℕ) < n → differentiable_on 𝕜 (iterated_deriv_within m f s) s) :=
+by simp only [times_cont_diff_on_iff_continuous_on_differentiable_on hs,
+  iterated_fderiv_within_eq_equiv_comp, continuous_linear_equiv.comp_continuous_on_iff,
+  continuous_linear_equiv.comp_differentiable_on_iff]
+
+/-- The `n+1`-th iterated derivative within a set with unique derivatives can be obtained by
+differentiating the `n`-th iterated derivative. -/
+lemma iterated_deriv_within_succ {x : 𝕜} (hxs : unique_diff_within_at 𝕜 s x) :
```

Comment:
Should this also come with a `'`-version that swaps the order of `deriv_within` and `iterated_deriv_within n`? Just like there is `pow_succ` and `pow_succ'`?

Reply:
Done.